config.org
#+begin_src elisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src

* Doom Emacs
Place your private configuration here! Remember, you do not need to run
=doom sync= after modifying this file!


Some functionality uses this to identify you, e.g. GPG configuration, email
clients, file templates and snippets. It is optional.
#+begin_src emacs-lisp :tangle yes
(setq user-full-name "Yi Ming"
     user-mail-address "ofseed@foxmail.com")
#+end_src

Doom exposes five (optional) variables for controlling fonts in Doom:

- ~doom-font~ -- the primary font to use
- ~doom-variable-pitch-font~ -- a non-monospace font (where applicable)
- ~doom-big-font~ -- used for ~doom-big-font-mode~; use this for
  presentations or streaming.
- ~doom-unicode-font~ -- for unicode glyphs
- ~doom-serif-font~ -- for the ~fixed-pitch-serif~ face

See =C-h v doom-font= for documentation and more examples of what they
accept. For example:
#+begin_src emacs-lisp :tangle no
(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
     doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
#+end_src
If you or Emacs can't find your font, use =M-x describe-font= to look them
up, =M-x eval-region= to execute elisp code, and =M-x doom/reload-font= to
refresh your font settings. If Emacs still can't find your font, it likely
wasn't installed correctly. Font issues are rarely Doom issues!

#+begin_src emacs-lisp :tangle yes
(setq doom-font (font-spec :family "LXGW WenKai Mono" :size 14.0)
      doom-variable-pitch-font (font-spec :family "LXGW WenKai" :size 14.0))
#+end_src

There are two ways to load a theme. Both assume the theme is installed and
available. You can either set ~doom-theme~ or manually load a theme with the
~load-theme~ function. This is the default:
#+begin_src emacs-lisp :tangle yes
(setq doom-theme 'doom-horizon)
#+end_src

This determines the style of line numbers in effect. If set to ~nil~, line
numbers are disabled. For relative line numbers, set this to ~relative~.
#+begin_src emacs-lisp :tangle yes
(setq display-line-numbers-type nil)
#+end_src

If you use ~org~ and don't want your org files in the default location below,
change ~org-directory~. It must be set before org loads!
#+begin_src emacs-lisp :tangle yes
(setq org-directory "~/Documents/org/")
#+end_src


* Custom Packages
Whenever you reconfigure a package, make sure to wrap your config in an
~after!~ block, otherwise Doom's defaults may override your settings. E.g.
#+begin_src emacs-lisp :tangle no
  (after! PACKAGE
    (setq x y))
#+end_src
The exceptions to this rule:

  - Setting file/directory variables (like ~org-directory~)
  - Setting variables which explicitly tell you to set them before their
    package is loaded (see =C-h v VARIABLE= to look up their documentation).
  - Setting doom variables (which start with =doom-= or =+=).

Here are some additional functions/macros that will help you configure Doom.

- ~load!~ for loading external *.el files relative to this one
- ~use-package!~ for configuring packages
- ~after!~ for running code after a package has loaded
- ~add-load-path!~ for adding directories to the ~load-path~, relative to
  this file. Emacs searches the ~load-path~ when you load packages with
  ~require~ or ~use-package~.
- ~map!~ for binding new keys

To get information about any of these functions/macros, move the cursor over
the highlighted symbol at press =K= (non-evil users must press =C-c c k=).
This will open documentation for it, including demos of how they are used.
Alternatively, use =C-h o= to look up a symbol (functions, variables, faces,
etc).

You can also try =gd= (or =C-c c d=) to jump to their definition and see how
they are implemented.

Adjust emacs window size. Note that ~initial-frame-alist~ has no effect on
the window size of emacsclient.
#+begin_src emacs-lisp :tangle yes
(pushnew! default-frame-alist '(width . 100) '(height . 40))
#+end_src

Evil mode configuration, use =C-l= to clear search highlighting.
#+begin_src emacs-lisp :tangle yes
(use-package! evil
  :config
  (evil-global-set-key 'normal (kbd "C-l") 'evil-ex-nohighlight))
#+end_src

Emacs-Rime is used as a client of rime to provide the Chinese input method
inside Emacs. ~rime-lib-finalize~ is required to finalize librime, otherwise,
there will a segment fault be generated. Wrap it with a custom function,
since it will not be defined if emacs-rime is not loaded.
#+begin_src emacs-lisp :tangle yes
(use-package! rime
  :init (setq rime-show-candidate 'posframe)
  :custom (default-input-method "rime")
  :config
  (defun ym/rime-finalize()
    (when (fboundp 'rime-lib-finalize)
     (rime-lib-finalize)))
  :hook (kill-emacs . ym/rime-finalize))
#+end_src

Vterm is a terminal emulator inside Emacs based on libvterm.
The default shell is ~$SHELL~, set it to =fish=.
#+begin_src emacs-lisp :tangle yes
(use-package! vterm
  :custom (vterm-shell "fish"))
#+end_src

A websocket implementation in lisp, for emacs. Load websocket after org-roam
is loaded. This is a recommend configuration of org-roam-ui.
#+begin_src emacs-lisp :tangle yes
(use-package! websocket
  :after org-roam)
#+end_src

Preview org latex on startup and
Scale latex preview images, the default is 1.5.
#+begin_src emacs-lisp :tangle yes
(use-package! org
  :config
  (setq org-startup-with-latex-preview t
        org-format-latex-options (plist-put org-format-latex-options :scale 3.0)))
#+end_src

Org-Roam-UI is a frontend for exploring and interacting with the Org-Roam notes.
#+begin_src emacs-lisp :tangle yes
(use-package! org-roam-ui
  :after org-roam ;; or :after org
;; normally we'd recommend hooking orui after org-roam, but since org-roam does not have
;; hookable mode anymore, you're advised to pick something yourself
;; you don't care about startup time, use
;;:hook (after-init . org-roam-ui-mode)
  :config
  (setq org-roam-ui-sync-theme t
        org-roam-ui-follow t
        org-roam-ui-update-on-save t
        org-roam-ui-open-on-start t))
#+end_src

Org-Re-Reveal is used to export html files with reveal.js format.
Configure the highlight.js plugin of reveal.js to use the builtin monokai theme.
#+begin_src emacs-lisp :tangle yes
(use-package! org-re-reveal
  :config (setq org-re-reveal-highlight-css 'monokai))
#+end_src

Telega is a client for Telegram.
#+begin_src emacs-lisp :tangle yes
(use-package! telega
  :load-path "~/telega.el"
  :commands (telega)
  :defer t)
#+end_src


config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
(setq user-full-name "Yi Ming"
     user-mail-address "ofseed@foxmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

(setq doom-font (font-spec :family "LXGW WenKai Mono" :size 18))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-horizon)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/Documents/org/")


;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(pushnew! default-frame-alist '(width . 100) '(height . 40))

(use-package! evil
  :config
  (evil-global-set-key 'normal (kbd "C-l") 'evil-ex-nohighlight))

(use-package! rime
  :init (setq rime-show-candidate 'posframe)
  :custom (default-input-method "rime")
  :config
  (defun ym/rime-finalize()
    (when (fboundp 'rime-lib-finalize)
     (rime-lib-finalize)))
  :hook (kill-emacs . ym/rime-finalize))

(use-package! telega
  :load-path "~/telega.el"
  :commands (telega)
  :defer t)
